.github/actions/setup-build/action.yml

```diff
@@ -0,0 +1,511 @@
+    # differs from the one on macOS Intel (x86-64).
+    - name: Set up statically linked libraries (macOS)
+      if: runner.os == 'macOS' && steps.setup.outputs.deploy == 'true'
+      shell: bash
+      run: |
+        ### Set up statically linked libraries ###
+
+        mkdir static-lib
+        if [ "${{ runner.arch }}" == "ARM64" ]; then
+          brew_dir=/opt/homebrew
+          # Include directories, not located in the usual places,
+          # must be explicitly appended to the include paths.
+          export CPATH="$brew_dir/opt/gmp/include:$brew_dir/opt/mpfr/include:/opt/homebrew/opt/zstd/include${CPATH:+":$CPATH"}"
+          echo "CPATH=$CPATH" >>"$GITHUB_ENV"
+        else
+          brew_dir=/usr/local
+        fi
+        ln -s $brew_dir/opt/gmp/lib/libgmp.a static-lib/libgmp.a
+        ln -s $brew_dir/opt/mpfr/lib/libmpfr.a static-lib/libmpfr.a
+        ln -s $brew_dir/opt/zstd/lib/libzstd.a static-lib/libzstd.a
+        export LIBRARY_PATH="$(pwd)/static-lib${LIBRARY_PATH:+":$LIBRARY_PATH"}"
+        echo "LIBRARY_PATH=$LIBRARY_PATH" >>"$GITHUB_ENV"
+
+    - name: Get cache key for the image
+      if: steps.setup.outputs.flint == 'true'
+      shell: ${{ steps.setup.outputs.msys2 == 'true' && 'msys2 {0}' || 'bash' }}
+      run: |
+        ### Get cache key for the image ###
+
+        # See: https://github.com/actions/upload-artifact/issues/231
+        echo "ImageOS=$ImageOS" >>"$GITHUB_ENV"
+        echo "ImageVersion=$ImageVersion" >>"$GITHUB_ENV"
+
+    - name: Cache FLINT
+      id: cache-flint
+      if: steps.setup.outputs.flint == 'true'
+      uses: actions/cache@v4
+      with:
+        path: lib/flint
+        # Here, we set a conservative cache key so that the cache is not reused
+        # when the version or environment changes even slightly.
+        key: flint-${{ inputs.flint_version }}-${{ runner.os }}-${{ runner.arch }}-${{ env.ImageOS }}-${{ env.ImageVersion }}
+
+    # Unfortunately, the FLINT package lacks a static library (because building
+    # it is disabled by default) and is outdated. We build FLINT from source.
+    # Note that we always set --enable-static so that release and non-release
+    # builds can share the same cache.
+    - name: Build FLINT unless cached
+      if: steps.setup.outputs.flint == 'true' && steps.cache-flint.outputs.cache-hit != 'true'
+      shell: ${{ steps.setup.outputs.msys2 == 'true' && 'msys2 {0}' || 'bash' }}
+      run: |
+        ### Build FLINT ###
+
+        mkdir -p build/flint
+        cd build/flint
+        # We use the tarball from GitHub Releases.
+        # See: https://github.com/flintlib/flint/issues/2311
+        wget https://github.com/flintlib/flint/releases/download/v${{ inputs.flint_version }}/flint-${{ inputs.flint_version }}.tar.gz
+        tar -xf *.tar.gz
+        rm *.tar.gz
+        cd flint-*
+        ./configure --prefix=$GITHUB_WORKSPACE/lib/flint --enable-static
```

Comment:
It appear that the FLINT configure script automatically selects the CPU architecture to optimise performance, which results in non-portable binaries. To address this problem, the `--host` option (for `>= 3.2.0`) must be specified with the correct host triplet  (`$(gcc -dumpmachine)` may be enough). Homebrew uses the [this code](https://github.com/Homebrew/homebrew-core/blob/7d54986bb799e750bc2fcd61fb0977e929e9ae12/Formula/f/flint.rb#L51-L56).